src/java.base/share/classes/java/lang/invoke/LazyInitializingVarHandle.java

```diff
@@ -0,0 +1,117 @@
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package java.lang.invoke;
+
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+
+import java.util.Optional;
+
+import static java.lang.invoke.MethodHandleStatics.UNSAFE;
+import static java.lang.invoke.MethodHandleStatics.uncaughtException;
+import static java.lang.invoke.MethodHandles.Lookup.IMPL_LOOKUP;
+
+/**
+ * A lazy initializing var handle. It lazily initializes the referenced class before
+ * any invocation of the target var handle to prevent reading uninitialized static
+ * final field values.
+ */
+final class LazyInitializingVarHandle extends VarHandle {
+
+    private static final MethodHandle MH_ensureInitialized;
+    private final VarHandle target;
+    private final Class<?> refc;
+    private @Stable boolean initialized;
+
+    static {
+        try {
+            MH_ensureInitialized = IMPL_LOOKUP.findVirtual(LazyInitializingVarHandle.class, "ensureInitialized",
+                    MethodType.methodType(void.class));
+        } catch (Throwable ex) {
+            throw uncaughtException(ex);
+        }
+    }
+
+    LazyInitializingVarHandle(VarHandle target, Class<?> refc) {
+        super(target.vform, target.exact);
+        this.target = target;
+        this.refc = refc;
+    }
+
+    @Override
+    MethodType accessModeTypeUncached(AccessType at) {
+        return target.accessModeTypeUncached(at);
+    }
+
+    @Override
+    VarHandle asDirect() {
+        return target;
+    }
+
+    @Override
+    @ForceInline
+    boolean checkAccessModeThenIsDirect(AccessDescriptor ad) {
+        super.checkAccessModeThenIsDirect(ad);
+        return false;
+    }
+
+    @Override
+    public VarHandle withInvokeExactBehavior() {
+        if (!initialized && hasInvokeExactBehavior())
+            return this;
+        var updatedDelegate = target.withInvokeExactBehavior();
+        return initialized ? updatedDelegate : new LazyInitializingVarHandle(updatedDelegate, refc);
+    }
+
+    @Override
+    public VarHandle withInvokeBehavior() {
+        if (!initialized && !hasInvokeExactBehavior())
+            return this;
+        var updatedDelegate = target.withInvokeBehavior();
+        return initialized ? updatedDelegate : new LazyInitializingVarHandle(updatedDelegate, refc);
+    }
+
+    @Override
+    public Optional<VarHandleDesc> describeConstable() {
+        return target.describeConstable();
+    }
+
+    private void ensureInitialized() {
+        UNSAFE.ensureClassInitialized(refc);
+        this.initialized = true;
+    }
```

Comment:
This should probably have a fast‑path when `this.initialized` is `true`, so that it can be better constant folded by the **JIT**:
```suggestion
    private void ensureInitialized() {
        if (this.initialized) {
            return;
        }
        UNSAFE.ensureClassInitialized(refc);
        this.initialized = true;
    }
```

Reply:
I think I should probably update the MH lambda form as well, to skip the ensure call entirely if JVM has not constant-fold the inefficient form yet.

Reply:
Indeed, updating lambda form and cached mh works: the average invocation performance becomes ~0.8ns/op, equal to the direct version, compared to the original ~1ms/op. See the updated benchmark results in the main PR body.